x-pack/plugins/security_solution/server/endpoint/migrations/turn_off_agent_policy_features.ts

```diff
@@ -0,0 +1,89 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import type { Logger, ElasticsearchClient } from '@kbn/core/server';
+import type { AgentPolicy } from '@kbn/fleet-plugin/common';
+import type { AuthenticatedUser } from '@kbn/security-plugin/common';
+import { AppFeatureSecurityKey } from '@kbn/security-solution-features/keys';
+import pMap from 'p-map';
+import type { EndpointInternalFleetServicesInterface } from '../services/fleet';
+import type { AppFeaturesService } from '../../lib/app_features_service/app_features_service';
+
+export const turnOffAgentPolicyFeatures = async (
+  esClient: ElasticsearchClient,
+  fleetServices: EndpointInternalFleetServicesInterface,
+  appFeaturesService: AppFeaturesService,
+  logger: Logger
+): Promise<void> => {
+  const log = logger.get('endpoint', 'agentPolicyFeatures');
+
+  if (appFeaturesService.isEnabled(AppFeatureSecurityKey.endpointAgentTamperProtection)) {
+    log.info(
+      `App feature [${AppFeatureSecurityKey.endpointAgentTamperProtection}] is enabled. Nothing to do!`
+    );
+
+    return;
+  }
+
+  log.info(
+    `App feature [${AppFeatureSecurityKey.endpointAgentTamperProtection}] is disabled. Checking endpoint integration policies for compliance`
+  );
+
+  const { agentPolicy: agentPolicyService, internalSoClient } = fleetServices;
+  const updates: AgentPolicy[] = [];
+  const messages: string[] = [];
+  const perPage = 1000;
+  let hasMoreData = true;
+  let total = 0;
+  let page = 1;
+
+  do {
+    const currentPage = page++;
+    const { items, total: totalPolicies } = await agentPolicyService.list(internalSoClient, {
+      page: currentPage,
+      kuery: 'ingest-agent-policies.is_protected: true',
+      perPage,
+    });
+
+    total = totalPolicies;
+    hasMoreData = currentPage * perPage < total;
+
+    for (const item of items) {
+      messages.push(
+        `Policy [${item.id}][${item.name}] updated to disable agent tamper protection.`
+      );
+
+      updates.push({ ...item, is_protected: false });
+    }
+  } while (hasMoreData);
+
+  if (updates.length > 0) {
+    logger.info(`Found ${updates.length} policies that need updates:\n${messages.join('\n')}`);
+    const policyUpdateErrors: Array<{ id: string; error: Error }> = [];
+    await pMap(updates, async (update) => {
+      try {
+        return await agentPolicyService.bumpRevision(internalSoClient, esClient, update.id, {
+          user: { username: 'elastic' } as AuthenticatedUser,
+          removeProtection: true,
+        });
+      } catch (error) {
+        policyUpdateErrors.push({ error, id: update.id });
+      }
+    });
```

Comment:
Agree that this is not efficient. I saw it being done this way in couple places though, [example](https://github.com/elastic/kibana/blob/63b86df054c1c608c82f922033786bdc5254755b/x-pack/plugins/fleet/server/services/agent_policy.ts#L1051-L1073). Gonna reach out to Fleet team and ask if there is a reason AgentPolicySrvc doesn't expose `bulkUpdate`.

Reply:
Agree about doing this in bulk so we don't have to update the SO one by one. That being said, there is no method on `agentPolicySerice` that supports `bulk` currently. If Fleet team agrees on, we can add a new one trying to mimic what the package policy service does for `bulkUpdate` here: https://github.com/elastic/kibana/blob/main/x-pack/plugins/fleet/server/services/package_policy.ts#L942

As this not only updates the SO and it does a lot of other stuff before doing the final update, it might be complex but worth if we can do it and speed this up.

Reply:
@paul-tavares @kpollich 
Do you think it would be better to add `bulkUpdate` in the current PR or create a new PR and issue just for that?